### PR TITLE
crossjoins converted to inner join if parent filter has suitable join condition

### PR DESCRIPTION
**Pull request by max-hoffman**
Filter predicates can be pushed into Join nodes under certain conditions:
- Filter is the parent of a CrossJoin
- Predicate expressions cover the two Join node dependency trees

Filter predicates have to be divided specially to maintain logical equivalence. Only `AND` conjugates can be split and checked individually for join candidacy. If multiple conjugate trees match the join expression, a new conjugate `AND` of the subtrees are used as the join condition.

Nested CrossJoins should be converted to InnerJoins if the parent filter has suitable join comparisons for each.

### CODE REVIEW

File: sql/analyzer/optimization_rules.go [hunk 1]
```diff
@@ -159,6 +161,90 @@ func moveJoinConditionsToFilter(ctx *sql.Context, a *Analyzer, n sql.Node, scope
...
+	}
+	return le.Table() == lt.Name() && re.Table() == rt.Name() ||
+		le.Table() == rt.Name() && re.Table() == lt.Name()
+}
+
+// replaceCrossJoins replaces a filter nested cross join with an equivalent inner join
+// Shove the whole expression into the new inner join as long as at least one subexpression is
+// a valid join key.
+// Filter(e) -> CrossJoin(a,b) with InnerJoin(a,b,e)
+func replaceCrossJoins(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
```
Comment (zachmu) [inline sql/analyzer/optimization_rules.go:210]: Probably want this in its own file

**Review by zachmu — COMMENTED** [review]
Main issue is that this approach won't work with cross joins of 3 or more tables, where the tree structure is `filter(crossjoin(crossjoin(a,b), c))`

You need to work in both directions, top down and bottom up:

Get the filter predicates first

Then work bottom up replacing CrossJoin nodes to InnerJoin nodes as possible

Then finally remove the used predicates from the filter

See pushdownfilter for an example, it's doing something similar

File: sql/analyzer/replace_cross_joins.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+// getFilterConjugates does a BFS to split a filter expression into
+// conjugate expression trees (or predicate leaves) that can be rearranged
+// without breaking logical equivalence.
+func getFilterConjugates(f *plan.Filter) []sql.Expression {
```
Comment (max-hoffman) [inline sql/analyzer/replace_cross_joins.go:26]: these should probably be called disjunctions, the more i think about it

**Review by zachmu — APPROVED** [review]
Looks good generally

Main comment is more engine tests, everything else is kind of nitpicky

File: sql/analyzer/replace_cross_joins.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	return expression.TraverseUp(c, func(expr sql.Expression) bool {
+		switch e := expr.(type) {
+		case expression.Comparer:
+			return comparisonSatisfiesJoinCondition(e, j)
+		}
+		return false
+	})
+}
+
+// replaceCrossJoins recursively replaces filter nested cross joins with equivalent inner joins.
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/replace_cross_joins.go:86]: is `a > 1 OR b > 2` a predicate still?

File: sql/analyzer/replace_cross_joins.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
+		switch f := n.(type) {
+		case *plan.Filter:
+			var disjuncts []sql.Expression
+			var coveringDisjuncts map[int]struct{}
+			newF, err := plan.TransformUp(f, func(n sql.Node) (sql.Node, error) {
+				switch j := n.(type) {
+				case *plan.CrossJoin:
+					if disjuncts == nil {
+						disjuncts = getFilterDisjunctions(f)
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/replace_cross_joins.go:105]: my thinking was that the values don't need to be initialized in most cases

[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/replace_cross_joins.go:105]: it only inits once

[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/replace_cross_joins.go:105]: i thought?

File: enginetest/query_plans.go [hunk 5]
```diff
@@ -477,6 +477,133 @@ var PlanTests = []QueryPlanTest{
 			"     └─ IndexedTableAccess(mytable on [mytable.i])\n" +
 			"",
 	},
+	{
+		Query: `SELECT a.* FROM mytable a, mytable b where a.i = b.s`,
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply enginetest/query_plans.go]: added matching explicit/CROSS JOIN tests for every implicit test. Added new enginetests with row outputs for all new implicit and explicit tests.

**Pull request merged**